File: dep/Binacle.ViPaq/ViPaqSerializer.Deserialize.cs
using System.IO.Compression;
using System.Numerics;
using System.Runtime.CompilerServices;
using Binacle.ViPaq.Abstractions;
using Binacle.ViPaq.Helpers;
using Binacle.ViPaq.Models;
using Version = Binacle.ViPaq.Models.Version;

namespace Binacle.ViPaq;

public static partial class ViPaqSerializer
{
	internal static (TBin, IList<TItem>) DeserializeInternal<TBin, TItem, T>(
		byte[] data
	)
		where T : struct, IBinaryInteger<T>, INumber<T>, IComparable<T>
		where TBin : IWithDimensions<T>, new()
		where TItem : IWithDimensions<T>, IWithCoordinates<T>, new()
	{
		if (data == null || data.Length < 1)
		{
			throw new ArgumentException("Data is invalid or empty.", nameof(data));
		}

		using var memoryStream = new MemoryStream(data);

		// Read the first byte (encoding info) before any decompression
		var firstByte = (byte)memoryStream.ReadByte();
		var encodingInfo = EncodingInfo.FromByte(firstByte);
		EncodingInfoHelper.ThrowOnInvalidEncodingInfo<T>(encodingInfo);

		// Determine if the data is compressed
		using var dataStream = GetDecodingDataStream(memoryStream, encodingInfo);

		using var protocolReader = new ProtocolReader<T>(dataStream);

		var numberOfItems = protocolReader.ReadUInt16();

		var bin = new TBin();
		protocolReader.ReadDimensions<TBin, T>(ref bin, encodingInfo.BinDimensionsBitSize);

		var items = new List<TItem>();
		for (int i = 0; i < numberOfItems; i++)
		{
			var item = new TItem();
			protocolReader.ReadDimensions<TItem, T>(ref item, encodingInfo.ItemDimensionsBitSize);
			protocolReader.ReadCoordinates<TItem, T>(ref item, encodingInfo.ItemCoordinatesBitSize);
			items.Add(item);
		}

		return (bin, items);
	}

	[MethodImpl(MethodImplOptions.AggressiveInlining)]
	private static Stream GetDecodingDataStream(MemoryStream stream, EncodingInfo encodingInfo)
	{
		if (encodingInfo.Version == Version.Uncompressed)
		{
			return stream;
		}

		if (encodingInfo.Version == Version.CompressedGzip)
		{
			return new GZipStream(stream, CompressionMode.Decompress);
		}

		throw new NotSupportedException($"Version {encodingInfo.Version} is not supported");
	}
}

File: packages/binacle-vipaq/vipaq-serializer.deserialize.ts
import {EncodingInfoHelper} from "./helpers/EncodingInfoHelper";
import {ProtocolReader} from "./ProtocolReader";
import {EncodingInfo} from "./models/EncodingInfo";
import {Version} from "./models/Version";
import Bin from "./models/Bin";
import Item from "./models/Item";
import DeserializedResult from "./models/DeserializedResult";



export async function deserialize(data: Uint8Array): Promise<DeserializedResult> {
	if (!data || data.length < 1) {
		throw new Error("Data is invalid or empty.");
	}

	// Read the first byte (encoding info) before any decompression
	const firstByte = data[0];
	const restOfData = data.slice(1);

	const encodingInfo = EncodingInfoHelper.fromByte(firstByte);
	// EncodingInfoHelper.ThrowOnInvalidEncodingInfo<T>(encodingInfo);

	const dataStream = await getDecodingDataStream(restOfData, encodingInfo);
	// Determine if the data is compressed
	const protocolReader = new ProtocolReader(dataStream);
	const numberOfItems = protocolReader.readUint16();

	// Create a minimal bin and items list so we return a valid DeserializedResult.
	// The original implementation reads dimensions and coordinates based on encodingInfo,
	// but those helpers are not implemented here yet. Returning placeholder objects
	// that consumers can inspect is safer than leaving the function without a return.
	const bin = new Bin();
	protocolReader.readDimensions(bin, encodingInfo.binDimensionsBitSize);

	const items: Item[] = [];
	for (let i = 0; i < numberOfItems; i++) {
		const item = new Item();
		protocolReader.readDimensions(item, encodingInfo.itemDimensionsBitSize);
		protocolReader.readCoordinates(item, encodingInfo.itemCoordinatesBitSize);

		items.push(item);
	}

	return new DeserializedResult(bin, items);
}

async function getDecodingDataStream(data: Uint8Array, encodingInfo: EncodingInfo): Promise<DataView<ArrayBuffer>> {
	if (encodingInfo.version == Version.Uncompressed) {
		return new DataView(data.buffer);
	}

	if (encodingInfo.version == Version.CompressedGzip) {
		const blob = new Blob([data.buffer]);
		const ds = new DecompressionStream('gzip');
		const decompressedStream = blob.stream().pipeThrough(ds);
		const reader = decompressedStream.getReader();
		const chunks: Uint8Array[] = [];
		let done = false;
		while (!done) {
			const {value, done: readerDone} = await reader.read();
			if (value) chunks.push(value);
			done = readerDone;
		}
		const decompressed = new Uint8Array(chunks.reduce((acc, chunk) => acc + chunk.length, 0));
		let offset = 0;
		for (const chunk of chunks) {
			decompressed.set(chunk, offset);
			offset += chunk.length;
		}
		return new DataView(decompressed.buffer);
	}

	throw new Error(`version ${encodingInfo.version} is not supported`);

}
